## Business Understanding

### Project Context

The short-term rental market in Nairobi has witnessed significant growth driven by digital platforms like Airbnb. As competition intensifies, both new and experienced hosts, investors, and property managers are looking for data-driven insights to inform pricing, property configuration, and guest engagement strategies.

This project leverages publicly available Airbnb datasets to uncover trends, patterns, and performance benchmarks for listings within Nairobi. The primary focus is to **understand what drives revenue, occupancy, and guest satisfaction**, and to identify how different listing characteristics impact these outcomes.

### Business Objectives

#### 1. Optimize Listing Performance

* Identify key features (room type, number of beds, superhost status, photo count, etc.) that correlate with high revenue, occupancy, and reviews.
* Provide actionable recommendations to hosts on how to configure or enhance their listings.

#### 2. Understand Guest Behavior

* Analyze review patterns to understand when and how guests engage with listings.
* Use review frequency and timing to infer guest satisfaction and listing popularity.

#### 3. Benchmark Competitors

* Help hosts compare their performance (e.g occupancy, revenue, reviews) to market averages and top-performing listings.
* Reveal what differentiates top-tier listings from underperformers.

#### 4. Support Strategic Pricing and Investment

* Inform pricing strategy by linking listing features to revenue and occupancy.
* Assist real estate investors in identifying the most profitable listing types or locations within Nairobi.

### Key Business Questions

| Category               | Business Question                                                                     |
| ---------------------- | ------------------------------------------------------------------------------------- |
| Performance         | What listing features are most associated with high revenue or occupancy?             |
| Host Strategy       | Do superhosts earn more or get better reviews? Are cohosted listings more successful? |
| Guest Sentiment     | Are listings with more reviews better rated or more frequently booked?                |
| Property Setup     | What room configurations (beds, baths, capacity) yield higher returns?                |
| Time Dynamics      | How does performance vary over time? Which listings show consistent growth?           |
| Content Impact      | Does the number of listing photos impact guest satisfaction or revenue?               |
| Market Segmentation | Can we group listings into meaningful performance tiers (e.g., premium vs budget)?    |

### Data Assets

| Dataset                       | Purpose                                                                                       |
| ----------------------------- | --------------------------------------------------------------------------------------------- |
| Listings Data              | Core listing metadata including room type, host info, ratings, revenue, and occupancy metrics |
| Reviews Data                | Historical guest reviews by date, volume, and listing ID                                      |

### Success Criteria

#### Business Success

* Stakeholders gain a clear, evidence-based view of the Nairobi Airbnb market.
* Hosts adopt insights to improve their performance metrics.
* Potential investors identify data-driven opportunities for market entry.

#### Project Success

* Creation of an integrated dataset combining listing and review dimensions.
* Generation of dashboards and visualizations answering all key business questions.
* Delivery of strategic recommendations rooted in analysis, not assumptions.

### Strategic Impact 

This project doesn’t just report the numbers—it transforms them into **competitive intelligence**. Whether you're a host trying to climb the rankings or an investor deciding where to allocate capital, these insights bridge the gap between **raw data and revenue decisions.**

## INITIAL DATA UNDERSTANDING (IDE)

Every dataset tells a story, but before we dive into narratives about Nairobi’s most bankable Airbnbs, we’re flipping through the table of contents. This phase is about getting comfortable with the data: what’s there, what’s missing, and what might blindside us if we don’t pay attention now.

We’re working with two core datasets:

* **Listings Data**: the DNA of each Nairobi Airbnb; room types, host info, pricing, ratings, occupancy, and revenue.
* **Reviews Data**: how guests have interacted with these listings over time; review volume, timing, and engagement signals.

### What’s happening:

* Importing key libraries: pandas, numpy, matplotlib, seaborn for slicing, dicing, and plotting.
* Peeking at the first few rows of each dataset to catch the vibe - naming conventions, structure, and potential early weirdness (misnamed columns, mixed datatypes).
* Checking the shape: Is this a lean 300-row insights machine, or are we running big data plays? Spoiler: Listings are small but rich; Reviews give us temporal depth.
* Pulling metadata: datatypes, null values, unique identifiers, this is where data whispers its secrets.
* Running basic stats: mean, mode, spread of numeric columns, and a glance at categorical value counts, laying the foundation for segmenting hosts, room types, and review behavior.

This is where trust is built, between us and the dataset.

Exploration done right is part instinct, part structure - this is both.
And now, we’re ready to start asking Nairobi’s Airbnbs the right questions.

In [30]:

# Core Libraries for Data Manipulation and Numerical Computation
import pandas as pd
import numpy as np

# Visualisation Libraries
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.graph_objects as go
import plotly.express as px
from plotly.subplots import make_subplots

# Render plotly graph on GitHub
# import plotly.io as pio
# pio.renderers.default = 'notebook+png'

# # Balance  size and resolution of plotly graphs
# pio.defaults.width = 1100
# pio.defaults.height = 450
# pio.defaults.scale = 1.5  

# Time Series Utlities
import calendar

# Database Interraction
import sqlite3

# Expression Evaluation
import ast

# Suppress Warnings
import warnings
warnings.filterwarnings("ignore")

### 1. LISTINGS DATASET

In [31]:
# Load dataset
listings_df = pd.read_csv("Nairobi Air BnBs\Listings Data.csv")

listings_df.head()

listing_id                          listing_name  \
0       75683        Kiloranhouse Apt Prime Bedroom   
1      471581       Located In a Serene Environment   
2      906958  Makena's Place Karen - Flamingo Room   
3     1237886                             Hob House   
4     1803821      Makena's Place Karen - All Rooms   

                listing_type     room_type  \
0       Private room in home  private_room   
1             Entire cottage   entire_home   
2    Private room in cottage  private_room   
3  Room in bed and breakfast    hotel_room   
4    Private room in cottage  private_room   

                                     cover_photo_url  photos_count  host_id  \
0  https://a0.muscache.com/im/pictures/5497476/24...            13    26997   
1  https://a0.muscache.com/im/pictures/6434524/bc...            37  2280941   
2  https://a0.muscache.com/im/pictures/68ecc57f-d...            29  4856316   
3  https://a0.muscache.com/im/pictures/cbdab7e1-f...             8  6748840   
4  https://a0.muscache.com/im/pictures/786899c5-b...            58  4856316   

  host_name cohosts  superhost  ...  l90d_occupancy  l90d_adjusted_occupancy  \
0    Adriel     NaN      False  ...             0.0                      0.0   
1     Bella     NaN       True  ...             0.0                      0.0   
2    Chichi     NaN       True  ...             0.2                      0.0   
3       NaN     NaN      False  ...             0.0                      0.0   
4    Chichi     NaN       True  ...             0.1                      0.0   

   l90d_revpar  l90d_revpar_native  l90d_adjusted_revpar  \
0          1.5               194.8                   0.0   
1          0.0                 0.0                   0.0   
2         10.9              1406.7                   0.0   
3          0.0                 0.0                   0.0   
4         12.5              1615.0                   0.0   

   l90d_adjusted_revpar_native l90d_reserved_days  l90d_blocked_days  \
0                          0.0                  3                  0   
1                          0.0                  0                 76   
2                          0.0                 18                  0   
3                          0.0                  0                  0   
4                          0.0                  9                  0   

  l90d_available_days  l90d_total_days  
0                  87               90  
1                  90               90  
2                  72               90  
3                  90               90  
4                  81               90  

[5 rows x 60 columns]

In [32]:
# Check shape of dataset
print(f"The dataset has {listings_df.shape[0]} rows and {listings_df.shape[1]} columns.")

The dataset has 300 rows and 60 columns.


In [33]:
# Get column names
listings_df.columns

Index(['listing_id', 'listing_name', 'listing_type', 'room_type',
       'cover_photo_url', 'photos_count', 'host_id', 'host_name', 'cohosts',
       'superhost', 'latitude', 'longitude', 'guests', 'bedrooms', 'beds',
       'baths', 'registration', 'amenities', 'instant_book', 'min_nights',
       'cancellation_policy', 'currency', 'cleaning_fee', 'extra_guest_fee',
       'num_reviews', 'rating_overall', 'rating_accuracy', 'rating_checkin',
       'rating_cleanliness', 'rating_communication', 'rating_location',
       'rating_value', 'ttm_revenue', 'ttm_revenue_native', 'ttm_avg_rate',
       'ttm_avg_rate_native', 'ttm_occupancy', 'ttm_adjusted_occupancy',
       'ttm_revpar', 'ttm_revpar_native', 'ttm_adjusted_revpar',
       'ttm_adjusted_revpar_native', 'ttm_reserved_days', 'ttm_blocked_days',
       'ttm_available_days', 'ttm_total_days', 'l90d_revenue',
       'l90d_revenue_native', 'l90d_avg_rate', 'l90d_avg_rate_native',
       'l90d_occupancy', 'l90d_adjusted_occupancy', 'l9

In [34]:
# Get metadata
listings_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 60 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   listing_id                   300 non-null    int64  
 1   listing_name                 300 non-null    object 
 2   listing_type                 300 non-null    object 
 3   room_type                    300 non-null    object 
 4   cover_photo_url              300 non-null    object 
 5   photos_count                 300 non-null    int64  
 6   host_id                      300 non-null    int64  
 7   host_name                    290 non-null    object 
 8   cohosts                      103 non-null    object 
 9   superhost                    300 non-null    bool   
 10  latitude                     300 non-null    float64
 11  longitude                    300 non-null    float64
 12  guests                       257 non-null    float64
 13  bedrooms            

In [35]:
# Get descriptive statistics of numerical columns
listings_df.describe().T

count          mean           std         min  \
listing_id                   300.0  2.552773e+07  1.082975e+07  75683.0000   
photos_count                 300.0  3.011000e+01  1.727036e+01      0.0000   
host_id                      300.0  1.094595e+08  9.256623e+07  26997.0000   
latitude                     300.0 -1.282850e+00  3.638098e-02     -1.3796   
longitude                    300.0  3.679496e+01  4.233261e-02     36.6694   
guests                       257.0  3.653696e+00  2.258819e+00      1.0000   
bedrooms                     250.0  2.044000e+00  1.383372e+00      1.0000   
beds                         298.0  2.255034e+00  2.262725e+00      1.0000   
baths                        300.0  1.488333e+00  1.227076e+00      0.0000   
amenities                      0.0           NaN           NaN         NaN   
min_nights                   298.0  2.258389e+00  3.757419e+00      1.0000   
cleaning_fee                 292.0  8.119863e+00  1.395363e+01      0.0000   
extra_guest_fee              274.0  3.605839e+00  1.114826e+01      0.0000   
num_reviews                  299.0  5.657525e+01  7.350572e+01      1.0000   
rating_overall               283.0  4.763074e+00  1.813036e-01      3.8000   
rating_accuracy              283.0  4.796113e+00  1.769093e-01      3.8000   
rating_checkin               283.0  4.865018e+00  1.220928e-01      4.4000   
rating_cleanliness           283.0  4.777739e+00  1.853667e-01      3.8000   
rating_communication         283.0  4.859717e+00  1.562047e-01      4.0000   
rating_location              283.0  4.829682e+00  1.761555e-01      3.6000   
rating_value                 283.0  4.762898e+00  1.856599e-01      3.4000   
ttm_revenue                  300.0  3.946213e+03  6.597077e+03    115.0000   
ttm_revenue_native           300.0  5.100951e+05  8.525026e+05  14858.0000   
ttm_avg_rate                 300.0  5.782167e+01  6.123182e+01      9.6000   
ttm_avg_rate_native          300.0  7.474014e+03  7.914808e+03   1245.6000   
ttm_occupancy                300.0  1.860000e-01  1.893591e-01      0.0000   
ttm_adjusted_occupancy       300.0  1.543333e-01  2.000365e-01      0.0000   
ttm_revpar                   300.0  1.081200e+01  1.807529e+01      0.3000   
ttm_revpar_native            300.0  1.397523e+03  2.335625e+03     40.7000   
ttm_adjusted_revpar          300.0  8.911000e+00  1.973102e+01      0.0000   
ttm_adjusted_revpar_native   300.0  1.151923e+03  2.549495e+03      0.0000   
ttm_reserved_days            300.0  6.970667e+01  6.696738e+01     10.0000   
ttm_blocked_days             300.0  4.516333e+01  6.000895e+01      0.0000   
ttm_available_days           300.0  2.952933e+02  6.696738e+01     67.0000   
ttm_total_days               300.0  3.650000e+02  0.000000e+00    365.0000   
l90d_revenue                 300.0  7.451967e+02  1.416586e+03      0.0000   
l90d_revenue_native          300.0  9.647705e+04  1.833640e+05      0.0000   
l90d_avg_rate                300.0  5.757233e+01  6.111204e+01      0.0000   
l90d_avg_rate_native         300.0  7.456948e+03  7.915108e+03      0.0000   
l90d_occupancy               300.0  1.340000e-01  1.919204e-01      0.0000   
l90d_adjusted_occupancy      300.0  5.700000e-02  1.709241e-01      0.0000   
l90d_revpar                  300.0  8.279333e+00  1.574047e+01      0.0000   
l90d_revpar_native           300.0  1.071967e+03  2.037377e+03      0.0000   
l90d_adjusted_revpar         300.0  3.528333e+00  1.446100e+01      0.0000   
l90d_adjusted_revpar_native  300.0  4.569790e+02  1.871190e+03      0.0000   
l90d_reserved_days           300.0  1.254000e+01  1.717746e+01      0.0000   
l90d_blocked_days            300.0  9.880000e+00  2.123618e+01      0.0000   
l90d_available_days          300.0  7.746000e+01  1.717746e+01      7.0000   
l90d_total_days              300.0  9.000000e+01  0.000000e+00     90.0000   

                                      25%           50%           75%  \
listing_id                   1.796217e+07  2.824189e+07

In [44]:
# Check data completeness.
# Duplicates
print("Duplicates:", listings_df.duplicated().sum())

# Check for missing values
null_counts = listings_df.isna().sum()
null_percentages = (null_counts / len(listings_df)) * 100

# Combine and filter only columns with missing values
missing_summary = pd.concat([
    null_counts.rename("Null Count"),
    null_percentages.round(2).rename("Null Percentage (%)")
], axis=1)

missing_summary = missing_summary[missing_summary["Null Count"] > 0]

print("\nMissing Values Overview:\n")
print(missing_summary)

Duplicates: 0

Missing Values Overview:

                      Null Count  Null Percentage (%)
host_name                     10                 3.33
cohosts                      197                65.67
guests                        43                14.33
bedrooms                      50                16.67
beds                           2                 0.67
registration                   9                 3.00
amenities                    300               100.00
instant_book                  26                 8.67
min_nights                     2                 0.67
cancellation_policy            2                 0.67
cleaning_fee                   8                 2.67
extra_guest_fee               26                 8.67
num_reviews                    1                 0.33
rating_overall                17                 5.67
rating_accuracy               17                 5.67
rating_checkin                17                 5.67
rating_cleanliness            17         

### 2. REVIEWS DATASET

In [37]:
# Load dataset
reviews_df = pd.read_csv("Nairobi Air BnBs\Reviews Data.csv")

reviews_df.head()

listing_id        date  num_reviews                              reviewers
0    26848700  2025-03-01            1                              423509914
1    13381106  2024-08-01            4  214060904,81464737,70099846,126731097
2    14535080  2024-11-01            1                              596414752
3    13381106  2025-05-01            4  369670554,462860774,17201826,10332598
4    37808882  2024-11-01            2                      4454461,356767713

In [38]:
# Get dataset shape
print(f"The dataset contains {reviews_df.shape[0]} rows and {reviews_df.shape[1]} columns.")

The dataset contains 917 rows and 4 columns.


In [39]:
# Get column names
reviews_df.columns

Index(['listing_id', 'date', 'num_reviews', 'reviewers'], dtype='object')

In [40]:
# Get metadata
reviews_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 917 entries, 0 to 916
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   listing_id   917 non-null    int64 
 1   date         917 non-null    object
 2   num_reviews  917 non-null    int64 
 3   reviewers    917 non-null    object
dtypes: int64(2), object(2)
memory usage: 28.8+ KB


In [41]:
# Get descriptive statistics of numerical columns
reviews_df.describe().T

count          mean           std      min         25%  \
listing_id   917.0  2.588971e+07  1.094185e+07  75683.0  18401756.0   
num_reviews  917.0  1.658670e+00  1.176630e+00      1.0         1.0   

                    50%         75%         max  
listing_id   28835961.0  34573425.0  39875956.0  
num_reviews         1.0         2.0        12.0

In [42]:
# Get descriptive statistics of categorical columns
reviews_df.describe(include = 'O').T

count unique         top freq
date        917     12  2024-07-01  102
reviewers   917    900   104220019    3

In [43]:
# Check data completeness.
# Duplicates
print("Duplicates:", reviews_df.duplicated().sum())

# Check for missing values
null_counts = reviews_df.isna().sum()
null_percentages = (null_counts / len(reviews_df)) * 100

print("\nMissing Values Overview:\n")
print(pd.concat([null_counts.rename("Null Count"), null_percentages.round(2).rename("Null Percentage (%)")], axis = 1))

Duplicates: 0

Missing Values Overview:

             Null Count  Null Percentage (%)
listing_id            0                  0.0
date                  0                  0.0
num_reviews           0                  0.0
reviewers             0                  0.0


## DATA UNDERSTANDING

### LISTINGS DATA

This dataset captures detailed attributes and performance metrics for Airbnb listings in Nairobi. Below are the key columns selected for analysis, along with their meanings:

- listing_id - Unique identifier for each listing; used for merges and tracking.
- listing_name - Title of the listing, useful for context or identifying specific cases.
- listing_type - Type of property (e.g apartment, house, studio); used for segmentation.
- room_type - Describes guest access (e.g Entire home, Private room); important for performance comparisons.

- host_id - Unique host identifier; helps identify hosts with multiple listings.
- cohosts - Indicates if the listing is managed by multiple people; may impact professionalism and service.
- superhost - Boolean indicating whether the host is a Superhost; often correlates with trust, ratings, and bookings.

- photos_count - Number of listing photos; proxy for how well the listing is presented visually.

- guests - Maximum guest capacity allowed for the listing.
- bedrooms - Number of bedrooms in the property.
- beds - Total number of beds provided.
- baths - Number of bathrooms (including partials, like 1.5).

- instant_book - Whether guests can book instantly without host approval.
- min_nights - Minimum number of nights required per stay.
- cancellation_policy - Host's cancellation rules (e.g Flexible, Moderate, Strict).

- currency - Currency in which prices and revenue are reported (e.g USD or KES).
- cleaning_fee - Additional one-time fee charged to guests for cleaning.
- extra_guest_fee - Extra charge per guest above a base number.

- num_reviews - Total number of guest reviews; a proxy for popularity and exposure.
- rating_overall - Average guest rating out of 5.
- rating_accuracy - Rating for how accurate the listing description is.
- rating_cleanliness - Rating for how clean the property is.
- rating_communication - Rating for how well the host communicated.
- rating_location - Rating for the location of the listing.
- rating_value - Rating for whether the listing felt worth the price.
- rating_checkin - Rating for the ease of checking in.

- ttm_revenue - Total revenue earned over the trailing 12 months.
- ttm_occupancy - Occupancy rate over the trailing 12 months (0 to 1).
- ttm_revpar - Revenue per available night (RevPAR) over the trailing 12 months.
- ttm_avg_rate - Average nightly rate over the trailing 12 months.
- ttm_reserved_days - Number of nights booked in the last 12 months.
- ttm_available_days - Number of nights the listing was available in the last 12 months.

- l90d_revenue - Revenue earned over the last 90 days.
- l90d_occupancy - Occupancy rate for the last 90 days.
- l90d_revpar - RevPAR for the last 90 days.
- l90d_avg_rate - Average nightly rate for the last 90 days.
- l90d_reserved_days - Number of nights booked in the last 90 days.
- l90d_available_days - Number of nights available in the last 90 days.

#### EARLY INSIGHTS & OBSERVATIONS

Before diving into modeling or deeper analysis, we took a high-level tour of the dataset to get a sense of its health, structure, and quirks. Here’s what stood out:

- **Shape & Scope**: The dataset contains **300 listings** and **60 columns**, covering everything from listing configuration and host details to 12-month and 90-day performance metrics.

- **No Duplicates**: There are **no duplicate rows**, which is a great start, every listing appears to be uniquely represented.

- **Missingness Isn’t Catastrophic**:
  - Most critical performance and pricing fields (like ttm_revenue, ttm_occupancy, ttm_avg_rate) are complete.
  - Fields like guests, bedrooms, and instant_book have some missing values (14–16%) but are manageable through imputation or flagging.
  - The amenities column is entirely missing, so it will be excluded from this analysis.

- **Photos Tell a Story**:
  - Listings have an average of **30 photos**, with a maximum of **129**, suggesting some hosts put significant effort into visual marketing.

- **Listings Have Room Variety**:
  - The average number of guests per listing is **~3.6**, with some listings accommodating up to **16 people**.
  - Bedrooms, beds, and baths vary widely, a signal of diverse property types, from studios to larger villas.

- **Pricing Insights**:
  - Average **TTM nightly rate** is around **KES 5,782**, but the standard deviation is large, indicating **price variability** across listings.
  - **Cleaning fees** and **extra guest fees** are often zero, but some go up to **KES 12,600 and KES 10,000** respectively - useful when analyzing total price perception.

- **Review Data Looks Healthy**:
  - Listings have, on average, **56 reviews**, with some up to **678**, showing a mix of new and seasoned properties.
  - Average overall rating is **4.76**, with high scores across individual categories (communication, location, etc.). Nairobi hosts are performing well in guest satisfaction.

- **Occupancy Seems Low**:
  - Average **TTM occupancy** is just **18.6%**, and **L90D occupancy** is even lower at **13.4%**.
  - This could reflect **seasonality**, **underutilization**, or **competition** in the market - a key focus for deeper analysis.

- **Performance Spread**:
  - Revenue and RevPAR figures show **huge variance**, with a few listings pulling in thousands (or tens of thousands) more than others. This hints at **high-performing outliers**, which may skew overall metrics but offer valuable benchmarking insights.

In summary: the dataset is structurally solid, diverse in property types, and rich in performance signals. Some columns will need light cleaning or imputation, and occupancy figures point to potential inefficiencies in the market, all strong leads for deeper exploration.

### REVIEWS DATASET

This dataset captures guest engagement with Nairobi Airbnb listings, focusing on review activity over time.

- listing_id - Unique identifier linking each review record to its associated Airbnb listing.
- date - Month and year of the review count (usually aggregated monthly); used to analyze review trends over time.
- num_reviews - Total number of reviews received by the listing in that specific month; a proxy for popularity and booking activity.
- reviewers - List of unique reviewer IDs for the month; useful for guest-level analysis or de-duplication if needed.

### OBSERVATION AND EARLY INSIGHTS

This dataset captures monthly review activity across Airbnb listings in Nairobi. Here's what we observed:

- **Structure**: The dataset has **917 rows** and **4 columns**, tracking reviews over time for multiple listings.

- **No Missing or Duplicate Data**: All columns are fully populated, and there are no duplicate rows.

- **Temporal Coverage**:
  - The date column spans **12 distinct months**, indicating monthly aggregation.
  - Most active review month is **July 2024**, with 102 entries.

- **Review Volume is Low but Varied**:
  - Each listing appears only once per month, and **the average monthly reviews per listing is ~1.66**, with a maximum of **12**.
  - The median (50%) value is just **1 review**, suggesting **many listings get minimal monthly feedback**, but a few are more active.

- **High Listing Diversity**:
  - The dataset includes **900 unique reviewer sets** and **917 unique listing-month combinations**, meaning **most entries are distinct**.

- **Reviewers Column**:
  - While each row lists reviewer IDs, this column likely serves **less analytical value** for this phase unless user-level behavior is needed.

In short: this dataset complements the listings data well - it adds a **temporal engagement layer**, useful for trend analysis, identifying high-performing listings over time, and studying seasonal review activity.

For this reason, we merge the two datasets